presto-main/src/main/java/com/facebook/presto/type/Re2JRegexp.java

```diff
@@ -0,0 +1,160 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.type;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.google.re2j.Matcher;
+import com.google.re2j.Options;
+import com.google.re2j.Pattern;
+import io.airlift.log.Logger;
+import io.airlift.slice.Slice;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.primitives.Ints.checkedCast;
+import static com.google.re2j.Options.Algorithm.DFA_FALLBACK_TO_NFA;
+import static java.lang.String.format;
+
+public final class Re2JRegexp
+{
+    private static final Logger log = Logger.get(Re2JRegexp.class);
+
+    private static final java.util.regex.Pattern DOT_STAR_PREFIX_SUFFIX_PATTERN = java.util.regex.Pattern.compile("^(\\.\\*\\??)?(.*?)(\\.\\*\\??)?$");
+    private static final int CORE_PATTERN_INDEX = 2;
+
+    public final int dfaStatesLimit;
+    public final int dfaRetries;
+
+    public final Pattern re2jPattern;
+    public final Pattern re2jPatternWithoutDotStartPrefixSuffix;
+
+    public Re2JRegexp(int dfaStatesLimit, int dfaRetries, Slice pattern)
+    {
+        this.dfaStatesLimit = dfaStatesLimit;
+        this.dfaRetries = dfaRetries;
+
+        Options options = Options.builder()
+                .setAlgorithm(DFA_FALLBACK_TO_NFA)
+                .setMaximumNumberOfDFAStates(dfaStatesLimit)
+                .setNumberOfDFARetries(dfaRetries)
+                .setEventsListener(new RE2JEventsListener())
+                .build();
+
+        String patternString = pattern.toStringUtf8();
+        re2jPattern = Pattern.compile(patternString, options);
+
+        // Remove .* prefix and suffix. This will give performance boost when using RE2J Pattern.find() function.
+        java.util.regex.Matcher dotStarPrefixSuffixMatcher = DOT_STAR_PREFIX_SUFFIX_PATTERN.matcher(patternString);
+        checkState(dotStarPrefixSuffixMatcher.matches());
+        String patternStringWithoutDotStartPrefixSuffix = dotStarPrefixSuffixMatcher.group(CORE_PATTERN_INDEX);
+
+        if (!patternStringWithoutDotStartPrefixSuffix.equals(patternString)) {
+            re2jPatternWithoutDotStartPrefixSuffix = Pattern.compile(patternStringWithoutDotStartPrefixSuffix, options);
+        }
+        else {
+            re2jPatternWithoutDotStartPrefixSuffix = re2jPattern;
+        }
+    }
+
+    public boolean matches(Slice source)
+    {
+        return re2jPatternWithoutDotStartPrefixSuffix.find(source);
```

Comment:
I have always had concerns about `re2jPatternWithoutDotStartPrefixSuffix`. Doing this just seems really wrong. Nevertheless, if this has to be done, I'll accept this compromise.

I read more about RE2 recently. According to STEP 4 of https://swtch.com/~rsc/regexp/regexp3.html, RE2 deals with this problem internally. RE2 understands that it can implement `PartialMatch` more efficiently when it doesn't need to answer where the match is. If this is indeed the design of RE2, I don't understand why RE2 can't solve this problem nicely inside.